File: solutions/c/binary-search/2/binary_search.c
#include "binary_search.h"

const int *binary_search(const int value, const int *const arr, const size_t length) {
    size_t begin = 0;
    size_t end = length;
    size_t mid;
    int cur;
    while (begin < end) {
        mid = (begin + end - 1) / 2;
        cur = arr[mid];
        if (cur == value) {
            return arr + mid;
        }
        if (cur < value) {
            begin = mid + 1;
        } else {
            end = mid;
        }
    }
    return NULL;
}

File: solutions/kotlin/binary-search/1/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    fun search(list: List<Int>, item: Int): Int {
        var left = 0
        var right = list.size
        while (left < right) {
            val mid = (left + right - 1) / 2
            if (list[mid] == item) {
                return mid
            } else if (list[mid] > item) {
                right = mid
            } else {
                left = mid + 1
            }
        }
        throw NoSuchElementException("$list does not have $item")
    }
}